src/wifi/mod.rs

```diff
@@ -1,43 +1,149 @@
+        const AP_SEC_PAIR_WEP104             = 0x00000002,
+        // TKIP is supported for pairwise/unicast encryption
+        const AP_SEC_PAIR_TKIP               = 0x00000004,
+        // AES/CCMP is supported for pairwise/unicast encryption
+        const AP_SEC_PAIR_CCMP               = 0x00000008,
+        // 40/64-bit WEP is supported for group/broadcast encryption
+        const AP_SEC_GROUP_WEP40             = 0x00000010,
+        // 104/128-bit WEP is supported for group/broadcast encryption
+        const AP_SEC_GROUP_WEP104            = 0x00000020,
+        // TKIP is supported for group/broadcast encryption
+        const AP_SEC_GROUP_TKIP              = 0x00000040,
+        // AES/CCMP is supported for group/broadcast encryption
+        const AP_SEC_GROUP_CCMP              = 0x00000080,
+        // WPA/RSN Pre-Shared Key encryption is supported
+        const AP_SEC_KEY_MGMT_PSK            = 0x00000100,
+        // 802.1x authentication and key management is supported
+        const AP_SEC_KEY_MGMT_802_1X         = 0x00000200,
+    }
+}
 
 
 /// Scans for Wi-Fi access points.
 ///
 /// # Examples
 ///
 /// ```
-/// let access_points = network_manager::wifi::scan().unwrap();
+/// use network_manager::manager;
+/// use network_manager::wifi;
+/// use network_manager::device;
+/// let manager = manager::new();
+/// let mut devices = device::list(&manager).unwrap();
+/// let i = devices.iter().position(|ref d| d.device_type == device::DeviceType::WiFi).unwrap();
+/// let device = &mut devices[i];
+/// let access_points = wifi::scan(&manager, device).unwrap();
 /// println!("{:?}", access_points);
 /// ```
-pub fn scan() -> Result<Vec<AccessPoint>, String> {
-    // Scan for access points
-
-    let ap1 = AccessPoint {
-        ssid: "ap1".to_string(),
-        signal: 60,
-        security: vec![Security::WEP],
-    };
-
-    let ap2 = AccessPoint {
-        ssid: "ap2".to_string(),
-        signal: 92,
-        security: vec![Security::WPA1, Security::WPA2],
-    };
-
-    Ok(vec![ap1, ap2])
+pub fn scan(manager: &NetworkManager, device: &Device) -> Result<Vec<AccessPoint>, String> {
+    let mut access_points = Vec::new();
+
+    if device.device_type == device::DeviceType::WiFi {
+        let paths = try!(manager.get_device_access_points(&device.path));
+
+        for path in paths {
+            if let Some(access_point) = try!(get_access_point(manager, &path)) {
+                access_points.push(access_point);
+            }
+        }
+    } else {
+        return Err("Not a WiFi device".to_string());
+    }
+
+    access_points.sort_by_key(|ap| ap.strength);
+    access_points.reverse();
+
+    Ok(access_points)
 }
 
 
-#[derive(Debug)]
-pub enum Security {
-    None,
-    WEP,
-    WPA1,
-    WPA2,
+fn get_access_point(manager: &NetworkManager,
+                    path: &String)
+                    -> Result<Option<AccessPoint>, String> {
+    if let Some(ssid) = manager.get_access_point_ssid(path) {
```

Comment:
Now I see. `get_access_point_security` is a separate function, since it invokes three `NetworkManager` methods and does additional bitflag calculations - all needed to return a `Security` instance. We do not need a separate function for getting the SSID, since we can use the one `NetworkManager` provides directly as no additional calculations are needed. On the other hand we can put `get_access_point_security` as a method of `NetworkManager`, but I prefer having it as external function, since it uses only public methods of `NetworkManager`, and we avoid adding too much stuff there.